src/shared/vscode/commands2.ts

```diff
@@ -365,44 +369,192 @@ interface CommandInfo<T extends Callback> {
      * (default: 5 minutes)
      */
     readonly telemetryThrottleMs?: number | false
-}
-
-// This debouncing is a temporary safe-guard to mitigate the risk of instrumenting all
-// commands with telemetry at once. The alternative is to do a gradual rollout although
-// it's not much safer in terms of weeding out problematic commands.
-const telemetryInfo = new Map<string, { startTime: number; debounceCount: number }>()
 
-function getInstrumenter(id: string, threshold: number, telemetryName?: MetricName) {
-    const currentTime = Date.now()
-    const info = telemetryInfo.get(id)
+    /**
+     * The indexes of args in `execute()` that will be used to determine
+     * the uniqueness of the metric `vscode_executeCommand`.
+     * 
+     * - By default, `vscode_executeCommand` is throttled for metrics of the
+     *   same `id`.
+     * - Defining arg indexes for this property, the values of the args will be combined
+     *   with the `id`. This ensures throttling only happens for the same `id` + args
+     *   instead.
+     * - Indexes starts at 0.
+     *
+     * @example
+     * const argIndexes: ArgIndexes = {
+     *     0: undefined // the value is not important
+     *     1: 'source' // the value is the "source" attribute of `vscode_executeCommand`
+     * }
+     * ...
+     * myCommand.execute('thisIsUsed', 'SOURCE', 'thisIsNotUsed')
+     * ...
+     * // vscode_executeCommand -> { command: 'myCommand' source: 'SOURCE' } 
+     */
+    readonly argIndexes?: ArgIndexes
+}
 
+/** 
+ * Used by {@link CommandInfo}, will indicate that the arg should be used
+ * as a telemetry attribute.
+ */
+export type ArgIndexes = { [index: number]: ArgIndexType }
+type ArgIndexType = MetricAttribute | undefined
+
+/** Supported `vscode_executeCommand` attributes */
+const MetricAttributes = {
+    source: 'source'
```

Comment:
what are we trying to model here? can the createKey and related things use the names provided by the `register()` call?

Reply:
> what are we trying to model here?

We've defined the field `source` for `vscode_executeCommand`. The design is for a Command to use "source" which results in the value of the arg to be set in the `source` field of the `vscode_executeCommand` metric.

> can the createKey and related things use the names provided by the register() call?

We can't since we need to use the actual value of the arg and not the name to differentiate the key.

Reply:
The names are provided by the register() call. What I don't understand is why we need this extra, explicit `MetricAttributes` type.